### PR TITLE
HPCC-24059 Additional SOAPCALL option to log text at end of soap request

### PR DESCRIPTION
**Pull request by afishbeck**
Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by afishbeck** [comment]
@ghalliday not sure what to do about the eclhelper interface changed test when it's intentional.

**Comment by richardkchapman** [comment]
This needs to target master

**Comment by afishbeck** [comment]
@richardkchapman Retargeted to master.

File: ecl/hql/hqlgram.y [hunk 1]
```diff
@@ -3768,6 +3768,12 @@ soapFlag
                             parser->normalizeExpression($3, type_string, false);
                             $$.setExpr(createExprAttribute(logAtom, $3.getExpr()), $1);
                         }
+    | TOK_LOG '(' expression ',' expression ')'
+                        {
+                            parser->normalizeExpression($3, type_string, false);
+                            parser->normalizeExpression($5, type_string, false);
+                            $$.setExpr(createExprAttribute(logAtom, createComma($3.getExpr(), $5.getExpr())), $1);
```
Comment (ghalliday) [inline ecl/hql/hqlgram.y:3775]: minor: Normally you would not have a comma inside an attribute.  i.e:

    createExprAttribute(logAtom, $3.getExpr(), $5.getExpr())

and then use unwindChildren() to extract them

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 2]
```diff
@@ -18198,10 +18213,16 @@ ABoundActivity * HqlCppTranslator::doBuildActivityHTTP(BuildCtx & ctx, IHqlExpre
...
 
-    if (logText)
+    if (logTextArray.length())
     {
         MemberFunction func(*this, instance->startctx, "virtual void getLogText(size32_t & __lenResult, char * & __result, const void * _left) override");
-        doBuildFunctionReturn(func.ctx, unknownStringType, logText);
+        doBuildFunctionReturn(func.ctx, unknownStringType, &logTextArray.item(0));
+    }
+
+    if (logTextArray.length()>1)
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlhtcpp.cpp:18222]: trivial: duplicate changes suggest these functions should be commoned up.  Not for this PR though.

**Review by ghalliday — CHANGES_REQUESTED** [review]
@afishbeck one fairly minor comment

**Comment by ghalliday** [comment]
@afishbeck you should also increment 
#define ACTIVITY_INTERFACE_VERSION      652
since the helper has changed.

**Comment by afishbeck** [comment]
@ghalliday I'll increment the ACTIVITY_INTERFACE_VERSION.  But just for discussion, since we're setting and checking a flag, that would seem to make it backward compatible.  The extra logging just wouldn't appear on older builds.  I imagine this is to prevent confusion with unexpected/missing behavior, although in this case it's fairly minor?

File: ecl/hql/hqlgram.y (hunk 1 above)
Comment (afishbeck) [reply ecl/hql/hqlgram.y:3775]: Gotcha.  I think I remember how a comma might be used now, it's been a while.  I'll fix it.

**Comment by ghalliday** [comment]
We're mainly concerned about whether old queries will run without recompiling on new systems - and increasing that number will not prevent that.  It will probably prevent new code running on old systems, but that isn't generally an issue.

File: ecl/hql/hqlgram.y (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlgram.y:3775]: The commas are mainly used for datasets.  At some point I may get rid of most/all of them and use initializer lists instead.

**Comment by afishbeck** [comment]
@ghalliday removed createComma, and updated ACTIVITY_INTERFACE_VERSION.

**Review by ghalliday — APPROVED** [review]

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**